pubsub/gcloud/aio/pubsub/subscriber.py

```diff
@@ -0,0 +1,229 @@
+                ack_queue.task_done()
+
+            while time_budget > 0:
+                start = time.perf_counter()
+                try:
+                    ack_id = await asyncio.wait_for(
+                        ack_queue.get(), timeout=time_budget)
+                    ack_ids.append(ack_id)
+                    ack_queue.task_done()
+                except asyncio.TimeoutError:
+                    break
+                time_budget -= (time.perf_counter() - start)
+            try:
+                await subscriber_client.acknowledge(subscription,
+                                                    ack_ids=ack_ids)
+            except Exception as e:
+                log.warning(
+                    'Ack request failed, better luck next batch', exc_info=e)
+                continue
+
+            if metrics_client:
+                metrics_client.histogram('pubsub.acker.batch', len(ack_ids))
+
+            ack_ids = []
+
+    async def _fire_callback(semaphore: asyncio.Semaphore,
+                             message: SubscriberMessage,
+                             callback: ApplicationHandler,
+                             ack_queue: 'asyncio.Queue[str]',
+                             metrics_client: Optional[MetricsAgent]
+                             ) -> None:
+        try:
+            start = time.perf_counter()
+            await callback(message)
+            await ack_queue.put(message.ack_id)
+            if metrics_client:
+                metrics_client.increment('pubsub.consumer.succeeded')
+                metrics_client.histogram('pubsub.consumer.latency.runtime',
+                                         time.perf_counter() - start)
+        except Exception:
+            log.exception('Application callback raised an exception')
+            if metrics_client:
+                metrics_client.increment('pubsub.consumer.failed')
+        finally:
+            semaphore.release()
+
+    async def consumer(
+            message_queue: MessageQueue,
+            callback: ApplicationHandler,
+            ack_queue: 'asyncio.Queue[str]',
+            ack_deadline_cache: AckDeadlineCache,
+            consumer_pool_size: int,
+            metrics_client: Optional[MetricsAgent]) -> None:
+        semaphore = asyncio.Semaphore(consumer_pool_size)
+        while True:
+            await semaphore.acquire()
+
+            message, pulled_at = await message_queue.get()
+
+            ack_deadline = await ack_deadline_cache.get()
+            if (time.perf_counter() - pulled_at) >= ack_deadline:
+                if metrics_client:
+                    metrics_client.increment('pubsub.consumer.failfast')
+                message_queue.task_done()
+                semaphore.release()
+                continue
+
+            if metrics_client:
+                metrics_client.histogram(
+                    'pubsub.consumer.latency.receive',
+                    time.time() - message.publish_time.timestamp())
```

Comment:
@allandialpad we were discussing timezones here?

I think I might be missing something since both timestamps should be in utc?

Reply:
We use this for egress and it hasn't been an issue (yet). Also I recall in google's pubsub source code they relied on time.time() as well. Sometimes we do get small negative values in the receive latency 😄